iptables/table.go

```diff
@@ -795,17 +789,25 @@ func (t *Table) readHashesAndRulesFrom(r io.ReadCloser) (hashes map[string][]str
 		}
 		hashes[chainName] = append(hashes[chainName], hash)
 
-		// Either we've explicitly inserted a rule into this chain or it's one of the top-level chains that we may have hooked.
-		// Store off the complete contents of the rules so that we can generate precise deletes later if we need to.
-		if _, ok := t.chainToInsertedRules[chainName]; t.topLevelChains.Contains(chainName) || ok {
+		// Not our chain so cache the full rule in case we need to generate deletes later on.
+		// After scanning the input, we prune any chains of full rules that do not contain inserts.
+		if !t.ourChainsRegexp.MatchString(chainName) {
 			fullRule := string(line)
 			rules[chainName] = append(rules[chainName], fullRule)
+			nonCalicoChainHasInserts[chainName] = true
 		}
 	}
 	if scanner.Err() != nil {
 		log.WithError(scanner.Err()).Error("Failed to read hashes from dataplane")
 		return nil, nil, scanner.Err()
 	}
+
+	// Remove full rules for the non-Calico chain if it does not have inserts.
+	for chainName, _ := range rules {
+		if hasInserts, ok := nonCalicoChainHasInserts[chainName]; ok && !hasInserts {
```

Comment:
Can this conditional ever be true?  I think `ok` will only be true if the loop above put something in the map. However, the code never puts `false` in the map.

One of Go's idioms is to make the "zero value" for a type useful.  To that end, Go's maps return the zero value if you look up an unknown key, rather than panicking.  The zero value for bools is `false` so that means you can implement a set like this without having to check for membership:
```
set := map[string]bool{}
set["a"] = true
set["b"] = true
if set["a"] /* true */ ...
if set["c"] /* false */ ...
```
Note: we also have a `Set` utility struct in libcalico-go:
```
set := set.New()
set.Add("a")
set.Add("b")
if set.Contains("a") /*true*/
if set.Contains("c") /*true*/
```
The `Set` struct is a little more space efficient because it avoids storing a bool for each entry in the set.